### PR TITLE
Implement sorting of smart contracts, address transactions

### PR DESCRIPTION
**Pull request by sl1depengwyn**
Part of #7794 

## Changelog

- Moved address to transactions functions from `Explorer.Chain` module to `Explorer.Chain.Transaction` module
- Moved verified smart contracts function from `Explorer.Chain` module to `Explorer.Chain.SmartContract`
- Moved `list_top_tokens` function from `Explorer.Chain` module to `Explorer.Chain.Token`
- Added general purpose module for sorting and pagination and implemented sorting for smart contracts and address transactions through this module as stated in #7794
- Refactor sorting of tokens using new `Explorer.SortingHelper` module
- Added two new query parameters for `api/v2/smart-contracts` similarly to `api/v2/tokens/` in https://github.com/blockscout/blockscout/pull/7895 endpoint
  - `sort` - possible values: `balance`, `txs_count`
  - `order` - possible values: `asc`, `desc`
- Added two new query parameters for `api/v2/:addresses_hash/transactions` similarly to `api/v2/tokens/` in https://github.com/blockscout/blockscout/pull/7895 endpoint
  - `sort` - possible values: `value`, `fee`
  - `order` - possible values: `asc`, `desc`
- Added new fields to paging_params in order to be able to paginate with sorting, latest frontend handles it fine
- Added param to `BlockScoutWeb.Chain.next_page_params` in order to be able to specify custom function that extracts options from params and handle collisions in `BlockScoutWeb.Chain.paging_params` with no pain as in https://github.com/blockscout/blockscout/pull/8634

## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [ ] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 1]
```diff
@@ -5,14 +5,14 @@ defmodule Explorer.Chain.Transaction do
 
   require Logger
 
-  import Ecto.Query, only: [from: 2, preload: 3, subquery: 1, where: 3]
+  # import Ecto.Query, only: [from: 2, preload: 3, subquery: 1, where: 3]
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/transaction.ex:8]: Do we need this comment?

File: apps/explorer/lib/explorer/chain/block/reward.ex [hunk 2]
```diff
@@ -279,4 +279,11 @@ defmodule Explorer.Chain.Block.Reward do
       query
     end
   end
+
+  @spec address_has_rewards?(Address.t()) :: boolean()
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/block/reward.ex:283]: Could you please doc as well?

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 3]
```diff
@@ -1111,4 +1113,380 @@ defmodule Explorer.Chain.Transaction do
...
+          boolean()
+        ) :: [__MODULE__.t()]
+  def address_to_transactions_without_rewards(address_hash, options, old_ui? \\ true) do
+    paging_options = Keyword.get(options, :paging_options, Chain.default_paging_options())
+
+    address_hash
+    |> address_to_transactions_tasks(options, old_ui?)
+    |> wait_for_address_transactions()
+    |> Enum.sort(compare_custom_sorting(Keyword.get(options, :sorting, [])))
+    # |> Enum.sort(&{&1.block_number, &1.index}, &>=/2)
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/transaction.ex:1274]: Do we need this commented line?

File: apps/explorer/lib/explorer/chain/transaction.ex
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/transaction.ex:1]: Could you please add spec/doc for all new public functions?

**Review by vbaranov — COMMENTED** [review]
Rebase is required.

**Review by nikitosing — CHANGES_REQUESTED** [review]

**Review by vbaranov — APPROVED** [review]

File: apps/explorer/lib/explorer/sorting_helper.ex [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+            (field(as(^binding), ^column) > ^value or
+               (field(as(^binding), ^column) == ^value and ^apply_next_column(next_column, key)))
+        )
+    end
+  end
+
+  defp page_by_column(key, {column, binding}, order, next_column) when order in ~w(desc desc_nulls_first)a do
+    case key[column] do
+      nil ->
+        apply_next_column(next_column, key)
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/sorting_helper.ex:213]: Why did you skip here `is_nil` check?

**Review by nikitosing — CHANGES_REQUESTED** [review]

File: apps/explorer/lib/explorer/sorting_helper.ex [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  defp sorting_with_defaults(sorting, default_sorting) when is_list(sorting) and is_list(default_sorting) do
+    (sorting ++ default_sorting)
+    |> Enum.uniq_by(fn
+      {_, field} -> field
+      {_, field, as} -> {field, as}
+      {:dynamic, key_name, _, _} -> key_name
+    end)
+  end
+
+  defp apply_as(sorting, query) do
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/sorting_helper.ex:45]: Mb rename it to `apply_sorting_inner` or `apply_sorting_to_query`?

File: apps/explorer/lib/explorer/sorting_helper.ex (hunk 4 above)
Comment (sl1depengwyn) [reply apps/explorer/lib/explorer/sorting_helper.ex]: Great catch! I decided to implement a solution to keep only one copy of this function, basically these clauses now defined in for comprehension

**Review by nikitosing — APPROVED** [review]

**Pull request merged**